pkg/scheduler/eventhandlers.go

```diff
@@ -203,18 +203,10 @@ func (sched *Scheduler) updatePodInCache(oldObj, newObj interface{}) {
 		klog.ErrorS(nil, "Cannot convert newObj to *v1.Pod", "newObj", newObj)
 		return
 	}
-
-	// A Pod delete event followed by an immediate Pod add event may be merged
-	// into a Pod update event. In this case, we should invalidate the old Pod, and
-	// then add the new Pod.
-	if oldPod.UID != newPod.UID {
-		sched.deletePodFromCache(oldObj)
-		sched.addPodToCache(newObj)
-		return
-	}
```

Comment:
I'd expect this PR to focus on eliminating the potential racing (very rare although). To ensure not cause regression, clear the old UID in podStates properly.

Reply:
> 

Thank you very much for your description.

Before submitting this PR, I have checked the related issue(https://github.com/kubernetes/kubernetes/issues/85726) and PR(https://github.com/kubernetes/kubernetes/pull/91126).

To clarify: I didn't delete the code, I just moved it to `schedulerCache.updatePod` (line 469) so that the cache can be accessed safely.

Let's go through the current code workflow (in this PR) with  `Update Event` :

- the **first** one is {old UID -> new UID} as the pod is recreated in master scheduler, while at the same time, the standby scheduler lost network connection to API Server. 
  ===> **unchanged**

- so when the standby scheduler becomes the new master, the pod update event would cache.go#UpdatePod() -> cache.updatePod(). In this case, **we use the oldPod's UID as the key to check weather the pod exists and delete the oldPod, while adding the newPod with newPod's UID [*newKey*].**
   ===> **changed: `podStates` will not hold the oldPod's UID but the newPod's UID as the key.**

- next the pod gets deleted, and the **second** pod update (b/c deletionTimeStamp is set) is delivered. This time, it's {new UID -> new UID}. And when it hits cache.go#UpdatePod(), **`podStates` has hold the newPod's UID as the key.**
  ===> **changed: we can handle this update correctly.** 

- then when the pod gets deleted physically, cache#RemovePod() gets called. **Cause podStates has hold the newPod's UID as the key, we can delete the newPod by the newPod's UID.** 
  ===> **changed: we can handle this delete correctly.**